In [1]:
import pandas as pd

A. Ознакомление с датасетом

В этой и последующих задачах мы будем работать с датасетом влияния COVID-19 на образование, психическое здоровье и распорядок дня учащихся. Ознакомиться с полной версией датасета можно по ссылке: https://www.kaggle.com/datasets/kunal28chaturvedi/covid19-and-its-impact-on-students.

Датасет доступен под названием 'COVID-19 SSR.csv'

В датасете содержатся такие переменные как время на занятия, самостоятельное обучение, просмотр страниц в социальных сетях в день, а также соц-демографические характеристики учащихся (возраст, регион проживания) и другие.

Вам необходимо вывести наиболее частый тип переменных встречаемый в датасете.

In [2]:
df=pd.read_csv("COVID-19 SSR.csv")

In [3]:
df.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1182 entries, 0 to 1181
Columns: 20 entries, Unnamed: 0 to What you miss the most
dtypes: float64(5), int64(3), object(12)
memory usage: 184.8+ KB


In [4]:
df.dtypes

Unnamed: 0                                                                              int64
ID                                                                                     object
Region of residence                                                                    object
Age of Subject                                                                          int64
Time spent on Online Class                                                            float64
Rating of Online Class experience                                                      object
Medium for online class                                                                object
Time spent on self study                                                              float64
Time spent on fitness                                                                 float64
Time spent on sleep                                                                   float64
Time spent on social media                                  

B. Изменяем капслок

Не всегда данные в одной колонке имеют одинаковое оформление. Например, одно значение написано буквами верхнего регистра, а другое - нижнего. Для приведения значений в категориальных переменных к одному виду существует набор функций.

title() - преобразовывает слова в формате title, где первое слово в строке начинается с символа верхнего регистра, а остальные приводятся к символам нижнего регистра (Например, "привет, студенты!" преобразуется в "Привет, студенты!").

В датасете'COVID-19 SSR.csv' в переменной 'Rating of Online Class experience' все значения записаны символами верхнего регистра (Например, "GOOD").

Вам необходимо привести значения переменной 'Rating of Online Class experience' к значению "GOOD" с помощью функции title(). Например, df('Название переменной').str.title().

В выдаче выведите пять первых значения переменной 'Rating of Online Class experience'.

Датасет доступен под названием 'COVID-19 SSR.csv'

In [5]:
print(df['Rating of Online Class experience'].str.title().head(5))


0         Good
1    Excellent
2    Very Poor
3    Very Poor
4         Good
Name: Rating of Online Class experience, dtype: object


C. Сколько спят учащиеся?

В нашем датасете есть переменная, указывающая на среднее количество сна в день ['Time spent on sleep']. Для здорового человека нормально спать 7-8 часов. Но учащиеся могут спать больше или меньше необходимого времени сна. То есть они могут спать например меньше 7 (6.9 часов) и больше 8 часов. Оптимальным количеством сна является 7-8 часов сна.

Вам необходимо вывести количество учащихся, которые спят больше или меньше оптимального времени для сна. Одним из способом решения этой переменной будут следующие шаги:

1. Воспользоваться циклом для создания новой переменной (название которой может быть 'Sleep') по переменной['Time spent on sleep'] , где необходимо прописать следующие условия. Если ['Time spent on sleep'] меньше 9, но при этом больше 6.9, то это нормальное количество сна ('normal'), в другом случае значение переменной будет 'not normal'.
2. Сделать фильтр по новой переменной, таким образом, чтобы мы оставили строки со значением сна'not normal'.
3. Вывести количество строк отфильтрованного датасета (например, с помощью функции 'len()').
Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [6]:
df_covid=pd.read_csv("COVID-19 Survey Student Responses.csv")

In [7]:
df_covid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1182 entries, 0 to 1181
Data columns (total 19 columns):
 #   Column                                                                              Non-Null Count  Dtype  
---  ------                                                                              --------------  -----  
 0   ID                                                                                  1182 non-null   object 
 1   Region of residence                                                                 1182 non-null   object 
 2   Age of Subject                                                                      1182 non-null   int64  
 3   Time spent on Online Class                                                          1182 non-null   float64
 4   Rating of Online Class experience                                                   1158 non-null   object 
 5   Medium for online class                                                             1131 non-null

In [8]:

df_covid=df_covid.assign(Sleep=['normal' if 6.9 < x < 9 else 'not normal' for x in df_covid['Time spent on sleep']])
len(df_covid[df_covid['Sleep']=='not normal'])

529

D. Часы в день

Мы обратим внимание на переменную 'Time spent on TV', которая содержит в себе не только числовые показатели, но и категориальные значения (object), такие как "N", "n", "No tv". Для анализа переменной необходимо изменить другие значения помимо числовых на "0". Далее после обработки датасета выведите тип переменной.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [9]:
df_covid = pd.read_csv("COVID-19 Survey Student Responses.csv")
df_covid=df_covid.assign(type_of_num=[type(x) for x in df_covid['Time spent on TV']])
df_covid['type_of_num'].value_counts()

type_of_num
<class 'str'>    1182
Name: count, dtype: int64

In [11]:
def isNumber(s):
    try:
        float(s) or int(s)
        return True
    except ValueError:
        return False
    
    
df_covid['Time spent on TV'] = df_covid['Time spent on TV'].apply(
    lambda x: x if isNumber(x) else '0').astype(float)
df_covid['Time spent on TV'].dtype
# df_covid['Time spent on TV'].value_counts()


dtype('float64')

E. Взаимосвязано ли количество сна и время, проведенное в социальных сетях?

Для ответа на данный вопрос мы будем работать с двумя переменными 'Time spent on sleep' и 'Time spent on social media'. Мы хотим выяснить связано ли недосыпание учащихся с временем, проведенным ими в социальных сетях.

В начале перекодируйте переменные на бинарные категории: 'normal' и 'not normal'. Нормальное количество сна для людей – больше 7 часов. Нормальное время, проведенное в сети – это меньше 2 часов в день. Перекодируйте переменные, а затем воспользуйтесь результатом Хи-теста и выведите результат Хи-квадрата.

Одним из способом решения задачи являются следующие шаги:

1. Воспользоваться циклом для создания новой переменной (название которой может быть 'Sleep') по переменной['Time spent on sleep'] , где необходимо прописать следующие условия. Если ['Time spent on sleep'] меньше 9, но при этом больше 6.9, то это нормальное количество сна ('normal'), в другом случае значение переменной будет 'not normal'.
2. Воспользоваться циклом для создания еще одной новой переменной (название которой может быть 'Media') по переменной['Time spent on social media'] , где необходимо прописать следующие условия. Если ['Time spent on social media'] меньше 2, то это нормальное количество времяпровождения в социальных сетей ('normal'), в другом случае значение переменной будет 'not normal'.
3. Воспользуйтесь функцией 'pd.crosstab' для подготовки данных к выдачи Хи-квадрата. crosstab() — метод, который помогает создать кросс-таблицу, показывает распределение данных по двум переменным. То есть вам необходимо применить метод  crosstab для двух, созданных вами, переменных (в качестве примера - это переменные 'Sleep' и 'Media').
4. Воспользуйтесь функцией 'chi2_contingency' из пакета 'stats' и выведите значение Хи-квадрата.
Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [16]:
from scipy.stats import chi2_contingency
df_covid = pd.read_csv("COVID-19 Survey Student Responses.csv")

df_covid = df_covid.assign(
    Sleep=['normal' if 7 <= x else 'not normal' for x in df_covid['Time spent on sleep']])
df_covid = df_covid.assign(
    Media=['normal' if x < 2 else 'not normal' for x in df_covid['Time spent on social media']])


my_table = pd.crosstab(df_covid['Sleep'], df_covid['Media'])

res=chi2_contingency(my_table)
res[0]


4.0632496096156885

F. Взаимосвязано ли количество сна и время, проведенное на в социальных сетях? 2.0

Мы хотим выяснить, связано ли недосыпание учащихся с временем, проведенным ими в социальных сетях.

В начале перекодируйте переменные на бинарные категории: 'normal' и 'not normal'. Нормальное количество сна для людей – больше 7 часов. Нормальное время, проведенное в сети –это меньше 2 часов в день. Перекодируйте переменные, а затем воспользуйтесь результатом Хи-теста и выведите результат Хи-квадрата.

Напишите аналогичный код из предыдущей задачи. (Рассчитайте Хи-квадрат для переменных 'Time spent on sleep' и 'Time spent on social media') и выведите результат p-value.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [18]:
res[1]

0.04382608418680551

G. Замена значений переменных с одних на другие

Для удобной работы категориальные переменные перекодируют в числовые. В этой задаче мы будем работать с бинарной переменной 'Health issue during lockdown'.

Перекодируйте значения переменной "YES" и "NO", на "1" и "0", соответственно.

В выдаче выведите количество значений по данной переменной.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [21]:
df_covid = pd.read_csv("COVID-19 Survey Student Responses.csv")
df_covid["Health issue during lockdown"].replace(['YES','NO'],['1','0'],inplace=True)
print(df_covid['Health issue during lockdown'].value_counts())


Health issue during lockdown
0    1021
1     161
Name: count, dtype: int64


H. Сколько читают учащиеся, справляясь со стрессом?

Чтобы понять находятся ли опеределенное слово, символ или другой показатель в значениях переменных, можно воспользоваться функцией contains(). Подробно об этой функции можно прочесть по ссылке https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.contains.html. В этой задаче вы будете работать с переменной 'Stress busters'. Выведите количество учащихся, которые используют книги в качестве профилактики борьбы со стрессом (встречаемость 'book' в значениях переменной).

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [74]:
df_covid['Stress busters'].str.contains('book',regex=False).value_counts()

False    1097
True       85
Name: Stress busters, dtype: int64

I. Самая популярная платформа среди социльных сетей

Определите наиболее популярную медиаплатформу в социальных сетях. (переменная 'Prefered social media platform')

Отфильтруйте пользователей, которые используют эту платформу для социальной сети и вычислите, в среднем сколько времени в день проводят в социальных медиа (переменная 'Time spent on social media') пользователи, которые предпочитают популярную социальную платформу. Результат округлите до сотых.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [29]:
pl=df_covid['Prefered social media platform'].value_counts().head(1).index.item()
round(df_covid[df_covid['Prefered social media platform'] == pl]['Time spent on social media'].mean(),2)


2.91

J. Кто проводит больше времени в социальных медиа?

В предыдущей задаче мы выяснили, что пользователи, которые предпочитают Инстаграм, используют социальные медиа в среднем около 2.91 часа.

Чтобы понять насколько пользователи Инстаграма отличаются от пользователей, отдающих предпочтения другой социальной сети, необходимо сделать подобную аналитику.

Определите платформу, в которой пользователи проводят больше всего времени в среднем.

Результат выведите через пробел. В начале укажите название предпочитаемой социальной сети, а затем среднее количество времени, затрачиваемое на пользование социальными сетями через пробел. Результат округлите до сотых.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [33]:
pl=df_covid['Prefered social media platform'].value_counts().index.tolist()
d={}
for item in pl:
  d[item]=round(df_covid[df_covid['Prefered social media platform']
           == item]['Time spent on social media'].mean(),2)
res=[[k,v] for k,v in sorted(d.items(),key=lambda x: x[1],reverse=True)]
print(res[0][0],res[0][1])


Talklife 10.0
